lib/ecto/changeset/relation.ex

```diff
@@ -179,8 +179,12 @@ defmodule Ecto.Changeset.Relation do
                    &do_cast(relation, &1, &2))
   end
 
-  defp do_cast(%{related: model, on_cast: fun} = meta, params, nil) do
-    {:ok, apply(model, fun, [meta.__struct__.build(meta), params])
+  defp do_cast(%{related: model, on_cast: fun} = meta, params, nil) when is_function(fun) do
```

Comment:
I wonder if it would be helpful to check that the arity is 2, and if it isn't raise a helpful message explaining what the function should look like and that the arity should be 2?